### ISSUE TITLE
unable to retrieve value on post

### ISSUE THREAD

**Issue by shrirammuthuvezhappan** [Mastercard/terraform-provider-restapi#98]
I have a working curl statement which retrieves a value from from api on post. In the curl statement ( post )  I pass on some key value pair which in turn return an appropriate value. This working curl statement I want to do it using the mastercard plugin.

My working curl statement looks like this -
curl --insecure -X POST 'https://url.fqdn/get_hostname' --header 'Content-Type: application/json' --header 'Authorization: Basic tokenvalue' -d '{"key": "value","key": "value","key": "value"}'

Converting this to code - 

```
provider "restapi" {

    uri = "https://url.fqdn/"
    insecure = true 
    headers = {
      "Content-Type" = "application/json"
      "Authorization" = "Basic tokenvalue"
    }
    write_returns_object = true
}
resource "restapi_object" "get_hostname" {
   path = "/get_hostname"
   data = "{ \"key\": \"value\", \"key\": \"value\", \"key\": \"value\" }"
}
```

I have two questions 
1. How to get values on POST using this plugin
2. Where would the output value be stored once the execution happens, is there a property i can output

**Comment by DRuggeri** [comment]
Hi, @shrirammuthuvezhappan. Please do note that this provider is intended to manage persistent resources on an API server. So, with that in mind, if the POST is intended to create an object and you would like to manage it through it's whole lifecycle, great!

To get ahold of the data, there are two fields available to you in the provider today. The `api_data` field contains a parsed and typed representation of the object that was created while the `api_response` field contains the exact text that was returned by the API. These fields can then be used as inputs or variables to other Terraform objects.

Head on over to the README file for more details

**Comment by shrirammuthuvezhappan** [comment]
Hi DRuggeri , thanks for your response. 
The post statement ( in curl )  it is not creating anything, but it acts as a carrier for json data in the body. When the api call (post) hits our API, based on the values in the json body, the api returns a **string value**. In this case, using api_response field would I be able to read the returned string ( not JSON ) ?

I also wonder since nothing was actually created and the return type is string , I wonder what will be captured in api_data.

For some reason it doesn't seem to work, I get the following error message whenever I try to do  a terraform apply ( terraform plan goes through fine )

```
restapi_object.get_hostname: Creating...

Error: Internal validation failed. Object ID is not set, but *may* have been created. This should never happen!

  on main.tf line 13, in resource "restapi_object" "get_hostname":
  13: resource "restapi_object" "get_hostname" {
```

**Comment by DRuggeri** [comment]
Thanks, @shrirammuthuvezhappan. I think your explanation is able to confirm that the use case doesn't match the provider's use case. The provider is designed to make POST calls and read back the data of the resource created on the server side. Since you are just reading data from an API call, it sounds like you would probably be better served by the `external` provider.

**Comment by shrirammuthuvezhappan** [comment]
Thanks so much, once again for your response. I have one more query on a similar context and I think it could be related this one.

In the past few days I had been trying to do a POST using Mastercard API which creates an entry, for example - create an entry in our cmdb which returns http 200.

For this use case the Mastercard API works perfectly, I am able to create the entry in our CMDB.
But I still get the error message 

```
Error: Internal validation failed. Object ID is not set, but *may* have been created. This should never happen!

  on main.tf line 14, in resource "restapi_object" "create_hostname":
  14: resource "restapi_object" "create_hostname" {

```

As the error message says, the object is getting created I can confirm that , _But I do not understand _Error: Internal validation failed. Object ID is not set**, **This should never happen!__

Please can you tell me know what am I missing or what is the background behind this error message
If I could get the http response back or some kind of success response, I can do some error handling

**Comment by DRuggeri** [comment]
Right - this error message is indicating that the request was submitted and accepted by the API, but the API did not respond with a JSON object with the attribute configured (`id` by default) identifying the ID of the object.

What this is saying is "I created an object under /get_hostname and I expect an ID to be returned such that manipulations to /get_hostname/{ID} will work." This is why the use case you have in mind doesn't fit the provider. The expectation is that the API is providing information to the provider permitting ongoing management of an object rather than this provider being used as an arbitrary REST client to make unrelated calls.

**Comment by shrirammuthuvezhappan** [comment]
Hi DRuggeri - We managed to use this API for our purpose. The only change we did was we made our API return a JSON object on Post, then I was able to fetch the returned value using the property api_data and api_response.

An another question ...

1. First time when do a terraform plan and terraform apply, the post action happens and our API returns JSON - All Good. As we know the end result is that the resources created and its return values are stored in the state file 

2. Second time when I do a terraform plan, the mastercard provider is trying to do a http GET and its trying to verify the current state file resource value with API, rather than just doing a post and updating the state file. 
`
Error: Unexpected response code '405': {"Message":"The requested resource does not support http method 'GET'."}`

Is this behaviour expected ?

I find it quite strange why the provider is performing a HTTP GET rather than just doing a post, is there a way to explicitly tell the provider only to post stuff

I have thought of couple of workarounds, but I would like to know why :)

**Comment by shrirammuthuvezhappan** [comment]
I was doing some debugging the way in which Mastercard API works and I found by default whatever mentioned in resource property path is used to auto populate the POST, GET, PUT and Delete methods. In my case I wanted the GET to point to some other api url, I was able to use the read_path property to point the api to another url. This solved my purpose and now I am able to use the Master card provider seamlessly.

My Post API call on first run and on subsequent runs the GET deciding whether it must do a POST or not work perfectly ! 

**I have found an answer to my question**
